### ISSUE TITLE
'errno' builtin addition 

### ISSUE THREAD

**Issue by tyroguru** [bpftrace/bpftrace#933]
It would be fabulous to be able to access a threads 'errno' value for when system calls have failed. Currently there is no easy mechanism available to access this. Possibly an 'errno' builtin?

**Comment by acj** [comment]
I've done a bit of preliminary work on this and would like to gather feedback before I go any further.

I'm assuming that the main use case is userspace calls, so something like this:

```
$ bpftrace -e 'uretprobe:...:func_that_makes_failing_syscall /retval < 0/ 
    { printf("errno is %d", errno); }'
Attaching 1 probe...
errno is 2
```

Getting portable access to errno seems tricky because it's a libc implementation detail, and the details vary between architectures. On modern Linux, `errno` is usually a macro that expands to `__errno_location` (or `__errno` on Android/bionic), which is a symbol that bpftrace can attach to if we know where to find libc and if we don't mind attaching to an internal symbol.

For example, given this testprog `errno.c`:

```c
#include <errno.h>
#include <stdio.h>

int main()
{
  FILE* pf = fopen ("does/not/exist", "rb");
  (void)errno;
  return 0;
}
```

... this works for me on Alpine (and on Ubuntu if I attach to the symbol in `/lib/x86_64-linux-gnu/libc.so.6` instead):

```
$ gcc -o errno errno.c
$ bpftrace -e 'uretprobe:/lib/ld-musl-x86_64.so.1:__errno_location /comm == "errno"/ 
    { printf("[%s, %d] *%d = %d\n", comm, tid, retval, *retval); }' -c ./errno
Attaching 1 probe...
[errno, 9278] *593710468 = 0
[errno, 9278] *593710468 = 0
[errno, 9278] *593710468 = 2
```

Perhaps an `errno` builtin, when present, could attach to `__errno_location` behind the scenes and track errno values (or the memory address) in a map keyed by thread ID, and then look up the appropriate value when the builtin is evaluated. However, I could see this creating a lot of overhead for some workloads, and I'm not sure that we want a builtin quietly attaching a probe to something that's called frequently.

Another path might be to use the FS register on x86_64 (GS on i386, possibly TPIDR_EL0 on aarch64, unsure about others) in codegen to get access to thread-local storage, and then walk through the pthread struct to read the offset for errno. I haven't tested this, but it seems much more efficient than using a probe in the background.

Thoughts? Maybe there's a better way?

**Comment by danobi** [comment]
I’m nervous about reaching into implementation defined behavior. I guess I have to wonder what the use case is. IIUC the majority use case of errno can be accomplished by looking at syscall return value, which is just -errno, and is completely reliable.

**Comment by acj** [comment]
It feels like a convenience thing to me. Maybe you're debugging and know that a syscall has failed recently, but you don't want to type out a second probe for it and try to correlate the events. It might be useful for one-off scenarios like that.

@tyroguru, what did you have in mind?

**Comment by jordalgo** [comment]
Closing due to inactivity.

**Comment by jordalgo** [comment]
Well it seems like I need something like this now because `__sys_bpf` returns -1 and sets the errno for PROG_LOAD and MAP_CREATE

**Comment by danobi** [comment]
@jordalgo I'm confused - isn't `errno` a libc thing? I'm looking at `__sys_bpf` in the kernel and I see it returning things like `-EFAULT`

**Comment by jordalgo** [comment]
@danobi From the bpf man page:

"For a successful call, the return value depends on the operation:

       BPF_MAP_CREATE
              The new file descriptor associated with the eBPF map.

       BPF_PROG_LOAD
              The new file descriptor associated with the eBPF program.

       All other commands
              Zero.

       On error, -1 is returned, and [errno](https://man7.org/linux/man-pages/man3/errno.3.html) is set to indicate the error."

https://man7.org/linux/man-pages/man2/bpf.2.html

Locally, I'm seeing -1 returned for BPF_MAP_CREATE, BPF_PROG_LOAD

**Comment by danobi** [comment]
The man page had me doubting reality, so I ran a test:

```
$ gcc main.c

$ ./a.out
raw syscall failed, ret=-1, errno=0
wrapped syscall failed, ret=-1, errno=1

$ sudo ./a.out
raw syscall failed, ret=-7, errno=0
wrapped syscall failed, ret=-1, errno=7

$ pastes < main.c
https://pastes.dxuuu.xyz/vhu5bl
```

So I think this confirms `errno` is a libc abstraction -- libc is doing negative retval translation to errno.

**Comment by jordalgo** [comment]
So because libc is doing this translation and we only get -1 as the retval - how might we get the errno? (forgive my ignorance)

**Comment by danobi** [comment]
Yeah, so from userspace side we only get -1 as the retval. Meaning if you were to uprobe a syscall. But if you hook in syscall with kprobe/kfunc, then I think fidelity is preserved (just read retval). Unfortunately for libc defined things like `fopen`, then it is quite tricky to recover errno.

I think the method @acj described above works for both musl and glibc. POSIX defines errno quite vaguely (https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/errno.h.html):

> The <errno.h> header shall provide a declaration or definition for errno. The symbol errno shall expand to a modifiable lvalue of type int. It is unspecified whether errno is a macro or an identifier declared with external linkage

As defined in glibc (https://github.com/bminor/glibc/blob/db57da5be27a0887ce5131c071e2240f5c17aae1/stdlib/errno.h#L37):

```c
/* The error code set by various library functions.  */
extern int *__errno_location (void) __THROW __attribute_const__;
# define errno (*__errno_location ())
```

it's a function that returns the addr of a thread local. So hooking into that function ought to work. But given POSIX is vague, it's totally within a libc's right to change the impl. Meaning if we place `errno` as a builtin, it could break one day and we might be out of luck (depending on libc impl).

I think supporting some kind of way to access `errno` is valuable. But maybe wrong to place in language. Perhaps this would be a good candidate for user defined functions? Or some way to write libraries in bpftrace?

**Comment by jordalgo** [comment]
@danobi Thanks for the explanation! And you're right, while I was seeing a lot of -1 returns for `__sys_bpf` (when only printing if the retval was less than 0), I left the kretprobe attached for a while and was seeing other values e.g.
```
(BPF_PROG_LOAD, Operation not permitted)
(BPF_PROG_LOAD, Operation not permitted)
(BPF_PROG_LOAD, Operation not permitted)
(BPF_PROG_LOAD, Bad address)
(BPF_PROG_LOAD, Bad address)
(BPF_PROG_LOAD, Bad address)
```

**Comment by amscanne** [comment]
It seems like there's a consensus that we can't easily wrap the libc-defined `errno` behavior. I think it's the right decision to close this bug. However, maybe this is a good future blog post? Honestly `errno` is a bit of dark magic to most people, and it might be interesting just to explain how that works and tie it into the bpftrace use case (suggesting how to solve this problem by trapping `__errno_location`, or trapping on the raw system call). It's not necessarily something that you would have in official documentation, but good to have as part of the broader materials.

**Comment by amscanne** [comment]
Closing for now, if there are objections we can reopen.
